src/game_strings.cpp

```diff
@@ -0,0 +1,223 @@
+	std::string str = static_cast<std::string>(Get(params.string_id));
+
+	int num;
+	if (params.hex)
+		num = std::stoi(str, 0, 16);
+	else
+		num = std::stoi(str);
+
+	Main_Data::game_variables->Set(var_id, num);
+	return num;
+}
+
+int Game_Strings::GetLen(Str_Params params, int var_id) {
+	if (!ResizeWithId(params.string_id)) return -1;
+
+	int len = static_cast<std::string>(Get(params.string_id)).length();
+	Main_Data::game_variables->Set(var_id, len);
+	return len;
+}
+
+int Game_Strings::InStr(Str_Params params, std::string search, int var_id, int begin) {
+	if (!ResizeWithId(params.string_id)) return -1;
+
+	if (params.extract) {
+		search = static_cast<std::string>(Extract(static_cast<Str_t>(search), params.hex));
+	}
+
+	std::string str = static_cast<std::string>(Get(params.string_id));
+
+	Output::Debug("Searching for {} in {}", search, str);
+
+	int index = static_cast<std::string>(Get(params.string_id)).find(search, begin);
+	Main_Data::game_variables->Set(var_id, index);
+	return index;
+}
+
+int Game_Strings::Split(Str_Params params, std::string delimiter, int string_out_id, int var_id) {
+	if (!ResizeWithId(params.string_id)) return -1;
+
+	// always returns at least 1
+	int splits = 1;
+	size_t index = 0;
+	std::string str = static_cast<std::string>(Get(params.string_id));
+	std::string token;
+
+	params.string_id = string_out_id;
+
+	while (index = str.find(delimiter) != std::string::npos) {
+		token = str.substr(0, index);
+		Set(params, static_cast<Str_t>(token));
+		params.string_id++;
+		splits++;
+		str.erase(0, index + delimiter.length());
+	}
+
+	// set the remaining string
+	Set(params, static_cast<Str_t>(str));
+	Main_Data::game_variables->Set(var_id, splits);
+	return splits;
+}
+
+Game_Strings::Str_t Game_Strings::ToFile(Str_Params params, std::string filename, int encoding) {
+	std::string str = static_cast<std::string>(Get(params.string_id));
+
+	if (params.extract) {
+		filename = static_cast<std::string>(Extract(static_cast<Str_t>(filename), params.hex));
+	}
+
+	// this sucks but it is what maniacs does
+	filename = "Text/" + filename + ".txt";
+
+	auto txt_out = FileFinder::Game().OpenOutputStream(filename);
+	if (!txt_out) { txt_out = FileFinder::Save().OpenOutputStream(filename); }
+	if (!txt_out) {
+		Output::Warning("Maniac String Op toFile failed!");
+		return "";
+	}
+
+	txt_out << str;
+	txt_out.Close();
+
+	return static_cast<Game_Strings::Str_t>(str);
+}
+
+Game_Strings::Str_t Game_Strings::PopLine(Str_Params params, int offset, int string_out_id) {
+	if (!ResizeWithId(params.string_id)) return "";
+
+	int index;
+	std::string result;
+	std::string str = static_cast<std::string>(Get(params.string_id));
+
+	std::stringstream ss(str);
+
+	while (offset >= 0 && std::getline(ss, result)) { offset--; }
+
+	offset = ss.rdbuf()->in_avail();
+
+	Set(params, static_cast<Str_t>(ss.str().substr(str.length() - offset)));
+	params.string_id = string_out_id;
+	return Set(params, static_cast<Str_t>(result));
+}
```

Comment:
What kind of linebreak is Maniacs emitting when writing to files? Is it ``\r\n`` or ``\n``?

Can Maniacs process strings correctly that have both types of line breaks?  (also relevant for PopLine as there can be different line ending).

EasyRPG Player usually handles both types of line breaks independent of the platform. There is ``Utils::ReadLine`` that has the same interface as ``std::getline``.

--------

Also needs checking if Maniacs reads in binary mode or in text mode by just reading a binary file.

A good test here is a PNG image read into a string and then written to the drive again. If its still an image it must be binary mode.

PNG also has ``0x1A`` very early in the header which is "End Of File" in text mode so it will read 0-bytes right after it encountered ``PNG``.

Reply:
Ah, I did not know that about the `Utils` package -- I will check that out! As for line endings I will admit I did not test too extensively with return carriages -- will report back as I experiment with stuff.

Maniacs will only read files with a `txt` or `csv` extension when assigning to a string, so I renamed a `image.png` to `image.txt`. When reading the file as `sjis`/`ansi`, it read only the `PNG`, but reading it as `utf8` appeared to read nothing at all (perhaps it failed? not sure). 

I'm pretty dumb when it comes to all this encoding stuff so your input is extremely valuable here.